tactics/proof_global.ml

```diff
@@ -159,31 +163,24 @@ let return_proof { proof } =
      equations and so far there is no code in the CI that will
      actually call those and do a side-effect, TTBOMK *)
   (* EJGA: likely the right solution is to attach side effects to the first constant only? *)
-  let proofs = List.map (fun (c, _) -> (proof_opt c, eff)) initial_goals in
+  let proofs = List.map (fun (body, typ) -> (to_constr_body body, eff), to_constr_typ typ) initial_goals in
   proofs, Evd.evar_universe_context evd
 
 let close_proof ~opaque ~keep_body_ucst_separate ps =
-  let elist, uctx = return_proof ps in
+
   let { section_vars; proof; udecl; initial_euctx } = ps in
-  let { Proof.name; poly; entry; sigma } = Proof.data proof in
+  let { Proof.name; poly } = Proof.data proof in
+  let unsafe_typ = keep_body_ucst_separate && not poly in
+  let elist, uctx = prepare_proof ~unsafe_typ ps in
   let opaque = match opaque with Opaque -> true | Transparent -> false in
 
-  (* Because of dependent subgoals at the beginning of proofs, we could
-     have existential variables in the initial types of goals, we need to
-     normalise them for the kernel. *)
-  let subst_evar k = Evd.existential_opt_value0 sigma k in
-  let nf = UnivSubst.nf_evars_and_universes_opt_subst subst_evar (UState.subst uctx) in
+  let make_entry ((body, eff), typ) =
 
-  let make_entry (body, eff) (_, typ) =
     let allow_deferred =
-      not poly && (keep_body_ucst_separate ||
-                   not (Safe_typing.empty_private_constants = eff.Evd.seff_private))
+      not poly &&
+      (keep_body_ucst_separate
+       || not (Safe_typing.is_empty_private_constants eff.Evd.seff_private))
```

Comment:
Not sure I understand this condition on the side effects tho, it seems like reversed.

Reply:
See https://github.com/coq/coq/wiki/Notes-on-side-effects,-universes-and-proof-closing
>NB: if side_effects didn't put us in allow_deferred=true,
    we would declare no typ univs which is broken in the Qed case

However this does break Derive when combined with side effects https://github.com/coq/coq/issues/10890

Reply:
I see, thanks a lot for the pointer; as I have hinted in other comments I'm trying to make side effects to be per-proof, not per-entry; so that should help with the above problem and other hiccups in obligations too [tho there are other issues with sharing the evar_map there]